Makefile

```diff
@@ -126,80 +114,219 @@ acncli: acncli-binary acncli-archive
 
 # Build the Azure NPM binary.
-.PHONY: azure-npm-binary
 azure-npm-binary:
 	cd $(CNI_TELEMETRY_DIR) && CGO_ENABLED=0 go build -v -o $(NPM_BUILD_DIR)/azure-vnet-telemetry$(EXE_EXT) -ldflags "-X main.version=$(VERSION)" -gcflags="-dwarflocationlists=true"
 	cd $(NPM_DIR) && CGO_ENABLED=0 go build -v -o $(NPM_BUILD_DIR)/azure-npm$(EXE_EXT) -ldflags "-X main.version=$(VERSION) -X $(NPM_AI_PATH)=$(NPM_AI_ID)" -gcflags="-dwarflocationlists=true"
 
 
-########################### Container Images ########################
+##@ Containers
+
+CNI_IMAGE = azure-cni-manager
+CNS_IMAGE = azure-cns
+NPM_IMAGE = azure-npm
 
-# Build the tools images
-.PHONY: tools-images
-tools-images:
-	$(MKDIR) $(IMAGE_DIR)
-	docker build --no-cache -f ./tools/acncli/Dockerfile --build-arg VERSION=$(VERSION) -t $(AZURE_CNI_IMAGE):$(VERSION) .
-	docker save $(AZURE_CNI_IMAGE):$(VERSION) | gzip -c > $(IMAGE_DIR)/$(CNI_IMAGE_ARCHIVE_NAME)
+TAG               ?= $(VERSION)
+IMAGE_REGISTRY    ?= acnpublic.azurecr.io
+PLATFORM          ?= $(GOOS)/$(GOARCH)
+CONTAINER_BUILDER  = buildah
+
+# prefer buildah, if available, but fall back to docker if that binary is not in the path.
+ifeq (, $(shell which $(CONTAINER_BUILDER)))
+CONTAINER_BUILDER = docker
+endif
+
+## This section is for building individual container images.
+
+container-platform-tag: # util target to print the container tag
+	@echo $(subst /,-,$(PLATFORM))-$(TAG)
+
+containerize-buildah: # util target to build container images using buildah. do not invoke directly.
+	buildah build \
+		--jobs 16 \
+		--platform $(PLATFORM) \
+		-f $(DOCKERFILE) \
+		--build-arg VERSION=$(VERSION) $(EXTRA_BUILD_ARGS) \
+		-t $(REGISTRY)/$(IMAGE):$(TAG) \
+		.
+
+containerize-docker: # util target to build container images using docker buildx. do not invoke directly.
+	docker buildx build \
+		--platform $(PLATFORM) \
+		-f $(DOCKERFILE) \
+		--build-arg VERSION=$(VERSION) $(EXTRA_BUILD_ARGS) \
+		-t $(REGISTRY)/$(IMAGE):$(TAG) \
+		.
+
+container-tag-test: # util target to retag an image with -test suffix. do not invoke directly.
+	$(CONTAINER_BUILDER) tag \
+		$(REGISTRY)/$(IMAGE):$(TAG) \
+		$(REGISTRY)/$(IMAGE):$(TAG)-test
+
+container-push: # util target to publish container image. do not invoke directly.
+	$(CONTAINER_BUILDER) push \
+		$(REGISTRY)/$(IMAGE):$(TAG)
+
+container-pull: # util target to pull container image. do not invoke directly.
+	$(CONTAINER_BUILDER) pull \
+		$(REGISTRY)/$(IMAGE):$(TAG)
+
+container-info: # util target to write container info file. do not invoke directly.
+	# these commands need to be root due to some ongoing perms issues in the pipeline.
+	sudo mkdir -p $(IMAGE_DIR) 
+	sudo chown -R $$(whoami) $(IMAGE_DIR) 
+	sudo chmod -R 777 $(IMAGE_DIR)
+	echo $(IMAGE):$(TAG) > $(IMAGE_DIR)/$(FILE)
+
+cni-manager-image-name: # util target to print the CNI manager image name.
+	@echo $(CNI_IMAGE)
+
+cni-manager-image: ## build cni-manager container image.
+	$(MAKE) containerize-$(CONTAINER_BUILDER) \
```

Comment:
Oh I see, you get which it is, docker or buildah, and call it here!

Reply:
Hmm I tried to run this target and got back an error about not being able to find the output:
```
make cni-manager-image
make containerize-docker \
        PLATFORM=linux/amd64 \
        DOCKERFILE=tools/acncli/Dockerfile \
        REGISTRY=acnpublic.azurecr.io \
        IMAGE=azure-cni-manager \
        EXTRA_BUILD_ARGS='--build-arg PLATFORM=_' \
        TAG=v1.4.16-62-gaad8f7be3
make[1]: Entering directory '/root/go/src/github.com/Azure/azure-container-networking'
docker buildx build \
        --platform linux/amd64 \
        -f tools/acncli/Dockerfile \
        --build-arg VERSION=v1.4.16-62-gaad8f7be3 --build-arg PLATFORM=_ \
        -t acnpublic.azurecr.io/azure-cni-manager:v1.4.16-62-gaad8f7be3 \
        .
WARN[0000] No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
[+] Building 127.5s (16/17)
 => [internal] booting buildkit                                                                                                                                                                                                                                                   25.0s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                                                                                24.0s
 => => creating container buildx_buildkit_dazzling_vaughan0                                                                                                                                                                                                                        1.1s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                               0.0s
 => => transferring dockerfile: 713B                                                                                                                                                                                                                                               0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                  0.1s
 => => transferring context: 2B                                                                                                                                                                                                                                                    0.1s
 => [internal] load metadata for docker.io/library/golang:1.17                                                                                                                                                                                                                     4.2s
 => [build  1/10] FROM docker.io/library/golang:1.17@sha256:1a35cc2c5338409227c7293add327ebe42e1ee5465049f6c57c829588e3f8a39                                                                                                                                                      24.1s
 => => resolve docker.io/library/golang:1.17@sha256:1a35cc2c5338409227c7293add327ebe42e1ee5465049f6c57c829588e3f8a39                                                                                                                                                               0.0s
 => => sha256:8560fc463426dc7e494720250efec25cdae1c4bf796c1a0172f791c0c7dde1c6 156B / 156B                                                                                                                                                                                         0.1s
 => => sha256:c9cefb9872505d3a6fdcbbdbe4103393da3e384443c5a8cdd62bc368927ea1cc 134.83MB / 134.83MB                                                                                                                                                                                18.2s
 => => sha256:9297634c9537024497f76a2e1b374d8a315baa21d45bf36dc7980dc42ab93b0b 85.81MB / 85.81MB                                                                                                                                                                                  15.8s
 => => sha256:461bb1d8c517c7f9fc0f1df66c9dc34c85a23421c1e1c540b2e28cbb258e75f5 54.57MB / 54.57MB                                                                                                                                                                                  11.8s
 => => sha256:e6d3e61f7a504fa66d7275123969e9917570188650eb84b2280a726b996040f6 10.87MB / 10.87MB                                                                                                                                                                                   2.6s
 => => sha256:412caad352a3ecbb29c080379407ae0761e7b9b454f7239cbfd1d1da25e06b29 5.15MB / 5.15MB                                                                                                                                                                                     1.2s
 => => sha256:0c6b8ff8c37e92eb1ca65ed8917e818927d5bf318b6f18896049b5d9afc28343 54.92MB / 54.92MB                                                                                                                                                                                  10.5s
 => => extracting sha256:0c6b8ff8c37e92eb1ca65ed8917e818927d5bf318b6f18896049b5d9afc28343                                                                                                                                                                                          1.9s
 => => extracting sha256:412caad352a3ecbb29c080379407ae0761e7b9b454f7239cbfd1d1da25e06b29                                                                                                                                                                                          0.2s
 => => extracting sha256:e6d3e61f7a504fa66d7275123969e9917570188650eb84b2280a726b996040f6                                                                                                                                                                                          0.3s
 => => extracting sha256:461bb1d8c517c7f9fc0f1df66c9dc34c85a23421c1e1c540b2e28cbb258e75f5                                                                                                                                                                                          1.9s
 => => extracting sha256:9297634c9537024497f76a2e1b374d8a315baa21d45bf36dc7980dc42ab93b0b                                                                                                                                                                                          1.5s
 => => extracting sha256:c9cefb9872505d3a6fdcbbdbe4103393da3e384443c5a8cdd62bc368927ea1cc                                                                                                                                                                                          3.8s
 => => extracting sha256:8560fc463426dc7e494720250efec25cdae1c4bf796c1a0172f791c0c7dde1c6                                                                                                                                                                                          0.0s
 => [internal] load build context                                                                                                                                                                                                                                                 12.7s
 => => transferring context: 875.16MB                                                                                                                                                                                                                                             12.6s
 => [build  2/10] WORKDIR /go/src/github.com/Azure/azure-container-networking/                                                                                                                                                                                                     1.4s
 => [build  3/10] ADD . .                                                                                                                                                                                                                                                          1.5s
 => [build  4/10] RUN make all-binaries                                                                                                                                                                                                                                           61.0s
 => [build  5/10] RUN make acncli                                                                                                                                                                                                                                                  5.6s
 => [build  6/10] RUN rm -rf ./output/windows*                                                                                                                                                                                                                                     0.1s
 => [build  7/10] RUN rm -rf ./output/_/npm/*                                                                                                                                                                                                                                      0.2s
 => [build  8/10] RUN mv ./output /output                                                                                                                                                                                                                                          1.1s
 => [build  9/10] RUN find /output -name "*.zip" -type f -delete                                                                                                                                                                                                                   0.2s
 => [build 10/10] RUN find /output -name "*.tgz" -type f -delete                                                                                                                                                                                                                   0.2s
 => ERROR [stage-1 1/2] COPY --from=build /output/_/acncli/ .                                                                                                                                                                                                                      0.0s
------
 > [stage-1 1/2] COPY --from=build /output/_/acncli/ .:
------
Dockerfile:16
--------------------
  14 |     FROM scratch
  15 |     ARG PLATFORM
  16 | >>> COPY --from=build /output/${PLATFORM:-linux_amd64}/acncli/ .
  17 |     COPY --from=build /output /output
  18 |     ENV AZURE_CNI_OS=linux
--------------------
error: failed to solve: failed to compute cache key: failed to walk /tmp/buildkit-mount282207209/output/_: lstat /tmp/buildkit-mount282207209/output/_: no such file or directory
make[1]: *** [Makefile:185: containerize-docker] Error 1
make[1]: Leaving directory '/root/go/src/github.com/Azure/azure-container-networking'
make: *** [Makefile:216: cni-manager-image] Error 2
```

Reply:
😟 that's unexpected, I will try and repro. the pipeline runs this stage also...